tasks/mpi.py

```diff
@@ -0,0 +1,102 @@
+from copy import copy
+from os import environ
+from os.path import exists, join
+from subprocess import run
+
+from tasks.util.env import PROJ_ROOT, FAABRIC_INSTALL_PREFIX
+
+from invoke import task
+
+EXAMPLES_DIR = join(PROJ_ROOT, "examples")
+BUILD_DIR = join(EXAMPLES_DIR, "build")
+
+INCLUDE_DIR = "{}/include".format(FAABRIC_INSTALL_PREFIX)
+LIB_DIR = "{}/lib".format(FAABRIC_INSTALL_PREFIX)
+
+# As it appears in the docker-compose file
+FAABRIC_WORKER_NAME = "cli"
+MPI_DEFAULT_WORLD_SIZE = 1
+
+
+def _find_mpi_hosts():
+    """
+    Return all deployed MPI hosts. This is, active containers based on the
+    client image. This won't work if we change the container name used
+    to deploying replicas.
+    """
+    docker_cmd = "docker ps --format '{{.Names}}'"
+
+    active_containers = run(
+        docker_cmd, shell=True, check=True, capture_output=True
+    )
+
+    # Format output and filter only 'client' containers
+    mpi_hosts = [
+        c
+        for c in active_containers.stdout.decode("utf-8").split("\n")
+        if FAABRIC_WORKER_NAME in c
+    ]
+
+    return mpi_hosts
+
+
+# TODO: eventually move all MPI examples to ./examples/mpi
+@task
+def execute(ctx, example, clean=False, np=MPI_DEFAULT_WORLD_SIZE):
+    """
+    Runs an MPI example
+    """
+    exe_path = join("./examples/build/", example)
+
+    if not exists(exe_path):
+        raise RuntimeError("Did not find {} as expected".format(exe_path))
+
+    shell_env = copy(environ)
+    shell_env.update(
+        {
+            "LD_LIBRARY_PATH": LIB_DIR,
+        }
+    )
+
+    # Start up `np` faabric instances. Note that if there are any other
+    # scaled out processes they will be stopped and removed. Additionally,
+    # running /bin/cli.sh whilst having a scaled out cluster will stop & remove
+    # them as well.
+    scale_up_cmd = "docker-compose up -d --scale {}={} {}".format(
+        FAABRIC_WORKER_NAME,
+        np,
+        "--force-recreate" if clean else "--no-recreate",
+    )
```

Comment:
This file is quite awkward, and re-engineers some docker-compose functionality. I would have thought it would be easier to create a new docker-compose file for this native MPI stuff, with two services, one for the workers and one for the client. Then you could use vanilla `docker-compose` commands to control the workers directly (without having to do the filtering and scaling in here). You'd then need to create a similar script to `cli.sh` for connecting to the client.

Try to do as much with the `docker-compose` CLI and separate config file as you can, then see whether it's still necessary to have scripting around that, and perhaps just try a bash script or two.

Reply:
I agree with you on this one. The approach I've gone for in the end is the following:

1. Custom `docker-compose.yml` file in `docker/mpi-native-docker-compose.yml`: it has less dependencies than the original compose as it's not supposed to be used as a cli, rather just a service to run one-off functions.
2. Deployment scripted using `bin/run_mpi_native.sh` and feeding input variables via `docker/mpi-native.env`: using the `.env` file fits well within the `docker-compose` architecture and permits not handling the input variables in `bash` which is quite cumbersome (I am happy to support variables as CL arguments, but I'd rather not do so :grin: ).

One problem I've run with is that I've needed to build a separate docker image with a different entrypoint. This is because I need to compile the libraries and the examples at build time, so that I can set the corresponding MPI binary as entrypoint. Alternatively, I could:
* Modify the original `faabric` base image to build the `faabricmpi_native` target and the examples (which I'd rather not do).
* Script around `docker-compose` again to ensure that everything is in place before we bootstrap execution.

Right now, the build time is around 3' so it's bearable.